### DISCUSSION TITLE
Local and global model locations

### DISCUSSION THREAD

**Discussion by FabioNotaro2001** [NVIDIA/NVFlare#1154] [Q&A] [answered]
Hi everybody and thanks.

I'm working on a simple example of real world usage of NVFlare. In particular I have executed the example hello-pt-tb using 2 clients, a server, an admin and an overseer, correctly created by the provisioning process. 
Now I have two questions about the model and its update process.

1 This is a confirmation: is there a file that contains the global model? I only found the file cross_val_results.json, is it? If no, where can I find it?
2 During the training process I plotted the accuracy/loss using tensorboard, obtaining the result I attach here. But does this plot refer to the local model of the clients or to the global model? In particular, if possible, I'd like to have a plot of the accuracy on the global model, not on the local models.
![Schermata del 2022-11-14 11-51-21](https://user-images.githubusercontent.com/93600518/202423976-3d03c4cf-990f-45bf-a261-0c069aa7d337.png)

Thank you in advance for your availability and support. Sorry for any possible error or imprecision but I'm trying to learn this topic. I remain available for possible details request.

**Comment by chesterxgchen** [comment]
you can download the related job result (including model)  by using 
download_job <job_id>
from NVFLARE console

**Comment by YuanTingHsieh** [comment, marked as answer]
Hi @FabioNotaro2001 ,

The "config_fed_client.json" is for defining behavior on client sites.
In there we configure the "pt_learner" to do the local execution for each client site.
In the "pt_learner", we configure the writer to send loss and accuracies back to the server site.
(https://github.com/NVIDIA/NVFlare/blob/dev/examples/hello-pt-tb/app/custom/pt_learner.py#L170)

So the tensorboard you see, site-1 refers to the site-1's model validate using site-1's data.
site-2 refers to the site-2's model validate using site-2's data.

The "config_fed_server.json" is for defining bahavior on the server site.
In there we configure the "ScatterAndGather" with "PTFileModelPersistor".
So the global model weights will be stored in the server workspace.
As @chesterxgchen points out, users can use admin client console to download the job results from the server using command "download_job <job_id>"
The server side model should be inside "<workspace>/models/".
^ this answer your first question.


To plot the losses for the server/global model, there are several ways to do that.
I will just provide one easier way here:

1) Write your own ScatterAndGather based on our code. On server side, after aggregate, you need to add logic to validate the server/global model against server's data and then writing out to tensorboard. => This would require server site also has some data.

**Reply by FabioNotaro2001** [reply]
Hi, thanks for your support.

I cannot find the folder /models you cited, can you be more specific on how I can find it?

In addition, I'd like to have more information on the last point you wrote: how can I plot loss/accuracy for the global model? What do you mean with ScatterAndGather? I saw that's the name of an example, just like hello-pt-tb. Is it correct? What do you mean by saying "add logic... and writing out to tensorboard...this would require server site also has some data". In brief, sorry for my inexperience but could you be a little more detailed?
Thank you very much for your patience and availability

**Reply by chesterxgchen** [reply]
ScatterAndGather is special MPI compute patten , you can Google it

**Reply by chesterxgchen** [reply]
Flare has a build-in controller for the scatter and gather pattern. You might need read the basic communication architecture, you can find the info from NVFLARE documents. Once you understand the communication architecture, you will understand what should be done at server and what at client. And logic described by @YuanTingHsieh

**Reply by YuanTingHsieh** [reply]
For ScatterAndGather I mean this workflow we have in NVFlare: https://github.com/NVIDIA/NVFlare/blob/dev/nvflare/app_common/workflows/scatter_and_gather.py 
You can add new logic in it ^

You need to evaluate your global/server model using server-side data then can use something like SummaryWriter to write out the accuracy you want.
(please refer to here for more info on tensorboard saving: https://pytorch.org/docs/stable/tensorboard.html)
